### ISSUE TITLE
Multiple application generation using JDL issues and improvements

### ISSUE THREAD

**Issue by deepu105** [jhipster/jhipster-core#259]
From https://github.com/jhipster/generator-jhipster/pull/8335

- [x] Set default value for serviceDiscoveryType when app type gateway|microservice or when serviceDiscovery is enabled for the monolith

- [x] The below app doesn't compile since it provides wrong defaults for `devDatabaseType`, `prodDatabaseType`, `enableHibernateCache` and 
```
application {
  config {
    baseName store,
    applicationType microservice,
    authenticationType oauth2,
    databaseType mongodb
    searchEngine elasticsearch,
    serverPort 8082,
    packageName com.okta.developer.store
  }
  entities Product
}
entity Product
```

- [ ] If I don't specify searchEngine elasticsearch on my gateway, the UI is generated without a search field. It seems like this should be read from the microservice apps (which have elasticsearch specified) rather than requiring it on the gateway.

- [x]   `skipUserManagement` should be set as true for microservice by default
- [x]  microservices should not have these props in yo-rc  
    ```
     "skipServer": false,
    "useSass": false,
    "clientFramework": "angularX"
    ```
- [x] `useSass` should be true by default for monolith and gateway
- [x] Invalid errors in console when using microservice with mogodb and mysql together and generation fails (blocker)
```
Error: NoSQL entities don't have relationships.
ERROR! Error while parsing applications and entities from the JDL Error: NoSQL entities don't have relationships.
```

<details>
<summary>Here is the JDL</summary>
<pre>
application {
  config {
    baseName store,
    applicationType gateway,
    packageName com.jhipster.demo.store,
    serviceDiscoveryType eureka,
    authenticationType jwt,
    prodDatabaseType mysql,
    cacheProvider hazelcast,
    buildTool gradle,
    clientFramework react,
    useSass true,
    testFrameworks [protractor]
  }
  entities *
}

application {
  config {
    baseName invoice,
    applicationType microservice,
    packageName com.jhipster.demo.invoice,
    serviceDiscoveryType eureka,
    authenticationType jwt,
    prodDatabaseType mysql,
    buildTool gradle,
    serverPort 8081,
    skipUserManagement true
  }
  entities Invoice, Shipment
}

application {
  config {
    baseName notification,
    applicationType microservice,
    packageName com.jhipster.demo.notification,
    serviceDiscoveryType eureka,
    authenticationType jwt,
    databaseType mongodb,
    prodDatabaseType mongodb,
    devDatabaseType mongodb,
    cacheProvider no,
    enableHibernateCache false,
    buildTool gradle,
    serverPort 8082,
    skipUserManagement true
  }
  entities Notification
}

/* Entities for Store Gateway */

/** Product sold by the Online store */
entity Product {
    name String required
    description String
    price BigDecimal required min(0)
    size Size required
    image ImageBlob
}

enum Size {
    S, M, L, XL, XXL
}

entity ProductCategory {
    name String required
    description String
}

entity Customer {
    firstName String required
    lastName String required
    gender Gender required
    email String required pattern(/^[^@\s]+@[^@\s]+\.[^@\s]+$/)
    phone String required
    addressLine1 String required
    addressLine2 String
    city String required
    country String required
}

enum Gender {
    MALE, FEMALE, OTHER
}

entity ProductOrder {
    placedDate Instant required
    status OrderStatus required
    code String required
    invoiceId Long
}

enum OrderStatus {
    COMPLETED, PENDING, CANCELLED
}

entity OrderItem {
    quantity Integer required min(0)
    totalPrice BigDecimal required min(0)
    status OrderItemStatus required
}

enum OrderItemStatus {
    AVAILABLE, OUT_OF_STOCK, BACK_ORDER
}

relationship OneToOne {
    Customer{user(login) required} to User
}

relationship ManyToOne {
	OrderItem{product(name) required} to Product
}

relationship OneToMany {
   Customer{order} to ProductOrder{customer(email) required},
   ProductOrder{orderItem} to OrderItem{order(code) required} ,
   ProductCategory{product} to Product{productCategory(name)}
}

service Product, ProductCategory, Customer, ProductOrder, OrderItem with serviceClass
paginate Product, Customer, ProductOrder, OrderItem with pagination

/* Entities for Invoice microservice */
entity Invoice {
    code String required
    date Instant required
    details String
    status InvoiceStatus required
    paymentMethod PaymentMethod required
    paymentDate Instant required
    paymentAmount BigDecimal required
}

enum InvoiceStatus {
    PAID, ISSUED, CANCELLED
}

entity Shipment {
    trackingCode String
    date Instant required
    details String
}

enum PaymentMethod {
    CREDIT_CARD, CASH_ON_DELIVERY, PAYPAL
}

relationship OneToMany {
    Invoice{shipment} to Shipment{invoice(code) required}
}

service Invoice, Shipment with serviceClass
paginate Invoice, Shipment with pagination
microservice Invoice, Shipment with invoice

/* Entities for notification microservice */

entity Notification {
    date Instant required
    details String
    sentDate Instant required
    format NotificationType required
    userId Long required
    productId Long required
}

enum NotificationType {
    EMAIL, SMS, PARCEL
}

microservice Notification with notification

</pre>
</details>

- [x] json overwrite issue, described here https://github.com/jhipster/jhipster-core/commit/f97bb475af1b20360e4fa746cfd45788537f0f8a

**Comment by deepu105** [comment]
@MathieuAA done, will add here if I find anything more

**Comment by deepu105** [comment]
I think it does if its empty, but the generated json has wrong values already there

**Comment by deepu105** [comment]
I would suggest to leave out anything that generator already handles out of the json created by JH core so that we dont have to duplicate the logic, else move the logic to JHcore and reuse the same in generator for default values

**Comment by deepu105** [comment]
@MathieuAA the last issue is a blocker do you know any workarounds?

**Comment by MathieuAA** [comment]
Let me think about it

**Comment by MathieuAA** [comment]
It's quite easy in a JDL context:
  - The user writes one or more apps in one or more JDL file/s
  - If in any of them is a MS & has ES enabled, set it in the GW

Now, if the user decides to import them separately and not in the same time, the best way is ask during the GW creation if any of the MS has ES enabled

**Comment by deepu105** [comment]
@MathieuAA I was not asking about ES related one, look at the alst item in the issue list, there are 7 items, I upated few more. Its about a fatal error with a valid JDL

**Comment by MathieuAA** [comment]
Well, I'm surprised this still works as I've PRed the generator yesterday to allow MongoDB with relationships.  
However, without the PR in mind, I guess this was almost normal but should have been a warning in this case (as both are used, which is not possible in a monolith app, right?). 
The problem could still happen with cassandra & any SQL db

**Comment by MathieuAA** [comment]
What's the project? monolith or MS?

**Comment by deepu105** [comment]
@MathieuAA hey did you happen to figure out the last one?

**Comment by deepu105** [comment]
@MathieuAA seems like the last point is resolved in master, i'll let you confirm and tick it off

**Comment by MathieuAA** [comment]
I've yet to confirm it, but I suspected.

**Comment by deepu105** [comment]
i'll do a PR for the new issue I added at the end now

**Comment by MathieuAA** [comment]
Ticked the "blocker", the PR from last week fixed it.

**Comment by deepu105** [comment]
@MathieuAA was the 2 open item on this done? I lost track of it

**Comment by deepu105** [comment]
@MathieuAA the issue no. 2 here was working in last version but seems to be failing again, see below JDL for example
```
application {
  config {
    baseName notification,
    applicationType microservice,
    packageName com.jhipster.demo.notification,
    serviceDiscoveryType eureka,
    authenticationType jwt,
    databaseType mongodb,
    cacheProvider no,
    enableHibernateCache false,
    buildTool gradle,
    serverPort 8083,
    skipUserManagement true
  }
}
```

**Comment by MathieuAA** [comment]
@deepu105 I think I know why, gotta test.

> failing again

If it's failing, there must be an error somewhere. Is it a validation error?

**Comment by deepu105** [comment]
Let me try again. I might have been in between versions.

**Comment by pascalgrimaud** [comment]
I think it's related to:
- my PR https://github.com/jhipster/jhipster-core/pull/280 to fix https://github.com/jhipster/generator-jhipster/issues/8760
- and some improvements: https://github.com/jhipster/jhipster-core/commit/28cceaa43c4e863eaa284dc291bbd7241a52eb12

**Comment by MathieuAA** [comment]
I'll fix it right away

**Comment by MathieuAA** [comment]
@pascalgrimaud I don't think so, the default values aren't propagated. Either I forgot to do it, or there's a regression. I'd bet for the first option.

**Comment by deepu105** [comment]
@MathieuAA plz close this if you think its done

**Comment by MathieuAA** [comment]
There's the second point, I'm gonna close this one and open a new one for point .
